power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp

```diff
@@ -0,0 +1,471 @@
+        } catch (std::exception& e) {
+            handle_error(e);
+        }
+    }
+
+    std::string const& dataset_name() const { return dataset_->name; }
+
+    bool is_batch() const { return is_batch_; }
+
+    Idx batch_size() const { return batch_size_; }
+
+    Idx n_components() const { return static_cast<Idx>(buffers_.size()); }
+
+    Buffer const& get_buffer_info(Idx i) const { return buffers_[i]; }
+
+    void set_buffer(char const** components, void** data, Idx** indptrs) {
+        for (Idx i = 0; i != n_components(); ++i) {
+            auto const found = std::find_if(buffers_.begin(), buffers_.end(), [components, i](Buffer const& buffer) {
+                return buffer.component->name == components[i];
+            });
+            if (found == buffers_.end()) {
+                throw SerializationError{"Unkown component: " + std::string(components[i]) +
+                                         "! You need to supply the components which are present.\n"};
+            }
+            found->data = data[i];
+            if (found->elements_per_scenario < 0) {
+                found->indptr = {indptrs[i], (size_t)(batch_size_ + 1)};
+            }
+        }
+    }
+
+    void parse() const {
+        root_key_ = "data";
+        try {
+            for (Buffer const& buffer : buffers_) {
+                parse_component(buffer);
+            }
+        } catch (std::exception& e) {
+            handle_error(e);
+        }
+    }
+
+  private:
+    msgpack::object_handle handle_{};
+    std::string version_;
+    bool is_batch_{};
+    MetaDataset const* dataset_{};
+    std::map<std::string, std::vector<MetaAttribute const*>> attributes_;
+    Idx batch_size_{}; // for single dataset, the batch size is one
+    std::vector<Buffer> buffers_;
+    // attributes to track the movement of the position
+    // for error report purpose
+    mutable std::string_view root_key_;
+    mutable std::string_view component_key_;
+    mutable std::string_view attribute_key_;
+    mutable Idx scenario_number_{-1};
+    mutable Idx element_number_{-1};
+    mutable Idx attribute_number_{-1};
+
+    static std::vector<char> json_to_msgpack(std::string_view json_string) {
+        nlohmann::json const json_document = nlohmann::json::parse(json_string);
+        std::vector<char> msgpack_data;
+        nlohmann::json::to_msgpack(json_document, msgpack_data);
+        return msgpack_data;
+    }
+
+    static std::string_view key_to_string(msgpack::object_kv const& kv) {
+        try {
+            return kv.key.as<std::string_view>();
+        } catch (msgpack::type_error&) {
+            throw SerializationError{"Keys in the dictionary should always be a string!\n"};
+        }
+    }
+
+    void post_serialization() {
+        if (handle_.get().type != msgpack::type::MAP) {
+            throw SerializationError{"The root level object should be a dictionary!\n"};
+        }
+        get_value_from_root("version", msgpack::type::STR) >> version_;
```

Comment:
The whole serializer/deserializer would have code not very consistent with rest of the repository, just like `meta_data.hpp`. And the whole serializer/deserializer is about using `msgpack`. To make a separate wrapper header is for me difficult to reason.

Reply:
to me, the (de)serializer is independent of the format used - it might as well be XML or plain JSON. the fact that we choose msgpack is not a design decision but an implementation detail. in the same link you shared about the optional dependencies (i agree with you on that), they also comment on the dependencies are used (and how). usage of a function like `extract_attribute(serialization_type_tag, data, component)` could for instance be completely independent of msgpack, even though the function itself strongly serialization format-dependent.

in extension, I like the design of this class the way you have written it right now precisely because it's relatively easy to extract the msgpack implementation details away from this file

Reply:
I understand your point. On the other hand, I don't want to go for pre-mature abstraction. So let's leave it as is and abstract `msgpack` away to separate headers if we indeed later start to support another serialization format.

Reply:
ok for now, but i did want to start the discussion (for background: i am speaking from experience here. i've had to try to remove `opencv-3.4` from the codebase to upgrade to `opencv-4.x` and even just that basically required reimplementing everything even though only core functionality was used)